File: 155-min-stack/155-min-stack.cpp
#include <stack>

class MinStack {
public:
    stack<pair<int,int>> st;
    int min;
    
    MinStack() {
        min = INT_MAX;
    }
    
    void push(int val) {
        if (val < min) min = val;
        st.push(make_pair(val, min));
    }
    
    void pop() {
        auto popped = st.top();
        st.pop();
        if (popped.second == min && !st.empty()) min = st.top().second;
        else if (st.empty()) min = INT_MAX;
    }
    
    int top() {
        return st.top().first;
    }
    
    int getMin() {
        return min;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 155-min-stack/155-min-stack.py
class MinStack:
# Two stack approach:
# 1. One to store the numbers
# 2. One to store the current minimum
    def __init__(self):
        self.nums = []
        self.curr_min = [] # (curr_min, count)

    def push(self, val: int) -> None:
        self.nums.append(val)
        
        if not self.curr_min or val < self.curr_min[-1][0]: 
            self.curr_min.append([val, 1])
        elif val == self.curr_min[-1][0]:
            self.curr_min[-1][1]+=1

    def pop(self) -> None:
        popped = self.nums.pop()
        if self.curr_min and self.curr_min[-1][0] == popped: 
            self.curr_min[-1][1]-=1
        if self.curr_min[-1][1] == 0:
            self.curr_min.pop()

    def top(self) -> int:
        return self.nums[-1]
        

    def getMin(self) -> int:
        return self.curr_min[-1][0]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()